docs/source/error-codes/self-service/index.rst

```diff
@@ -7,6 +7,138 @@ Self-Service Error Codes (Experimental)
 .. toctree::
    :hidden:
 
+Overview
+*********
+
+
+TODO: Failure vs. error.
+
+.. _gRPC: https://grpc.github.io/grpc/core/md_doc_statuscodes.html
+
+Participant API server has been using a combination of gRPC_
+status codes and custom messages to signal erroneous or unexpected conditions.
+This approach remains unchanged in principle while self-service error codes aim at
+enhancing it by providing:
+
+- improved consistency of the returned errors across API endpoints,
+
+- richer message format with clearly distinguished machine readable parts to facilitate
+  automated error handling strategies,
+
+- complete inventory of all error codes with an explanation, suggested resolution and
+  more (as documented in this section).
+
+
+The goal is to enable users, developers and operators to act upon encountered
+errors in a self-service manner such that the need to escalate the problem
+or file a ticket occurs occurs infrequently.
+
+Feature flag
+************
+You can enable self-service error-codes by specifying ``--use-self-service-error-codes``
+from command line.
+
+By default self-service error codes are turned off.
+
+
+Anatomy of an error code
+*****************************
+
+TODO Error categories:
+
+ - what they are,
+
+ - how to make use of them,
+
+ - provide a list of all error categories.
+
+Error message
+===================
+.. code-block:: java
+
+    <ERROR_CODE_ID>(<CATEGORY_ID>,<CORRELATION_ID>:<HUMAN_READABLE_MESSAGE>
+
+The constituent parts are:
+
+  - ``<ERROR_CODE_ID>`` - a unique non empty string containing at most 63 characters:
+    upper-cased letters, underscores or digits.
+  - ``<CATEGORY_ID>`` - a small integer identifying the corresponding error categories.
+  - ``<CORRELATION_ID>`` - a string helpful at identifying originating request.
+    ``0`` if no correlation id is given. TODO: Are there limits to its size?
+     - The purpose of the correlation-id is to allow a user to clearly identify the request,
+       such that the operator can lookup any log information associated with this error.
+       TODO: Make sure that's true.
+  - ``:`` - a colon character that server as a separator for the machine and human readable parts.
+  - ``<HUMAN_READABLE_MESSAGE>`` - message targeting a human reader.
+
+A concrete example how a message might look like this:
+
+.. code-block:: java
+
+    TRANSACTION_NOT_FOUND(11,12345): Transaction not found, or not visible.
+
+Error metadata
+====================
+
+In addition error response contain the following metadata:
+- mandatory ``com.google.rpc.ErrorInfo`` containing <ERROR_CODE>,
+
+- mandatory ``com.google.rpc.RequestInfo`` containing trace id,
+
+- optional ``com.google.rpc.RetryInfo`` containing retry interval,
+
+- optional ``com.google.rpc.ResourceInfo`` containing key value pair
+  of the form: <RESOURCE_NAME>: <RESOURCE_IDENTIFIER>,
+  e.g. ""TRANSACTION_ID", "tId12345".
+  TODO: Document all possible resource names?
```

Comment:
Just say that the resources are implementation specific and may vary from ledger to ledger